### PR TITLE
Fix the issue that pod or container config file without metadata will…

### PR DESCRIPTION
**Pull request by teawater**
… crash containerd

Because RunPodSandbox and CreateContainer will access metadata
without check, pod or container config file without metadata will
crash containerd.

This patch add checks to handle the issue.

Fixes: #1009

Signed-off-by: Hui Zhu <teawater@hyper.sh>

### CODE REVIEW

File: pkg/server/container_create.go [hunk 1]
```diff
@@ -92,7 +92,11 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 	// Reserve the container name to avoid concurrent `CreateContainer` request creating
 	// the same container.
 	id := util.GenerateID()
-	name := makeContainerName(config.GetMetadata(), sandboxConfig.GetMetadata())
+	metadata := config.GetMetadata()
+	if metadata == nil {
+		return nil, errors.New("GetMetadata failed")
```
Comment (AkihiroSuda) [inline pkg/server/container_create.go]: nit: Go error should start with a lower character, e.g. `errors.New("failed to get metadata")`

Comment (mikebrow) [reply pkg/server/container_create.go]: /s/GetMetadata failed/container config must include metadata/

add a TODO to make container create work using image name if metadata name is not set..

File: pkg/server/sandbox_run.go [hunk 2]
```diff
@@ -58,7 +58,11 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 
 	// Generate unique id and name for the sandbox and reserve the name.
 	id := util.GenerateID()
-	name := makeSandboxName(config.GetMetadata())
+	metadata := config.GetMetadata()
+	if metadata == nil {
+		return nil, errors.New("GetMetadata failed")
```
Comment (mikebrow) [inline pkg/server/sandbox_run.go]: /s/GetMetadata failed/sandbox config must include metadata/

File: pkg/server/container_create.go [hunk 3]
```diff
@@ -91,8 +91,14 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 	// Generate unique id and name for the container and reserve the name.
 	// Reserve the container name to avoid concurrent `CreateContainer` request creating
 	// the same container.
+	// TODO(teawater): make container create work using image name if metadata name
```
Comment (Random-Liu) [inline pkg/server/container_create.go]: 1) I'm not sure whether we'd like to do that.
2) Even if we'd like to do that, we should do it in crictl instead of the cri plugin.

**Comment by Random-Liu** [comment]
Sorry for not mentioning this in the last pass.

Please add unit test cases for nil metadata. Just copy an existing unit test, change the metadata to nil, and assert an error instead of panic.

**Comment by Random-Liu** [comment]
nvm, we don't have unit test for `CreateContainer` and `RunPodSandbox`. Can you add a simple integration test for it? An integration test example: https://github.com/containerd/cri/blob/master/integration/container_log_test.go

**Comment by Random-Liu** [comment]
Let's get this in first. I'm fine with adding the integration test in another PR.
